lib/mapreduce/engine.go

```diff
@@ -135,33 +134,32 @@ func (e *engine) run(ctx context.Context, rng Range) <-chan error {
 		}
 		last = end
 
-		ready, next := chain.next()
+		link := chain.next()
 
-		go func() {
-			defer func() {
-				wg.Done()
-				if next != nil {
-					close(next)
-				}
-			}()
+		doMap := func(ctx context.Context) (interface{}, error) {
+			sem := pool.get()
+			defer sem.done()
 
 			rng := Range{
 				Start: start,
 				End:   end,
 			}
 
-			if err := pool.wait(ctx); err != nil {
-				errch <- err
-				return
+			if err := sem.wait(ctx); err != nil {
+				return nil, err
 			}
 
-			out, err := e.m.Map(ctx, rng)
-			if err != nil {
-				errch <- err
-				return
-			}
+			return e.m.Map(ctx, rng)
+		}
 
-			if err := pool.done(ctx); err != nil {
+		go func() {
+			defer func() {
+				link.done()
```

Comment:
If a mapper finishes and returns a `nil` output, then we will unlock a chain early. The ordered `link.done()` needs to always wait for the previous chain to finish.

This was a problem in the earlier design as well.